### PR TITLE
Fix tolerance handling for joint-based forwarding

### LINKED ISSUE #18: [forward_controllers] Using tolerances lead to assertion_error if RobotHW doesn't provide a full tolerance
When using tolerances with the passthrough controllers, the controller crashes with an assertion_error if the RobotHW doesn't provide a full error state including accelerations and efforts.

e.g. when using with the current implementation inside the `ur_robot_driver` the joint error looks like this:

```
positions[]
  positions[0]: 0
  positions[1]: 0
  positions[2]: 0
  positions[3]: 0
  positions[4]: 0
  positions[5]: 0
velocities[]
  velocities[0]: 0
  velocities[1]: 0
  velocities[2]: 0
  velocities[3]: 0
  velocities[4]: 0
  velocities[5]: 0
accelerations[]
effort[]
```

That they are all exactly 0 seems a bit suspicious, but that's a separate topic. However, that `accelerations` and `effort` are empty leads to the mentioned assertion_error from https://github.com/fzi-forschungszentrum-informatik/cartesian_ros_control/blob/0a908ce355e13b16f55b4633623f32c9d3906240/pass_through_controllers/include/pass_through_controllers/pass_through_controllers.hpp#L249-L251

Is that on purpose and the implementation in `ur_robot_driver` should be changed and if so, why is acceleration required, but not effort? This seems a bit arbitrary to me.

---
**Edit:** They are all 0 as I am using URSim for testing, where target == actual.

### PR DESCRIPTION
**Pull request by stefanscherzinger**
This changes the joint-based tolerance checking:

- Check each user-given tolerance field individually.
- Fail if either the tolerance is exceeded or if the robot driver does not provide semantically correct data (conservative fail).

Fixes #18

### CODE REVIEW

File: pass_through_controllers/include/pass_through_controllers/pass_through_controllers.hpp [hunk 1]
```diff
@@ -243,28 +243,40 @@ namespace trajectory_controllers {
...
-           std::abs(error.velocities[i]) > tolerances[i].velocity) ||
-          (tolerances[i].acceleration > 0.0 &&
-           std::abs(error.accelerations[i]) > tolerances[i].acceleration))
+      if (tolerances[i].position > 0.0)
+      {
+        if (error.positions.size() == tolerances.size())
+        {
+          return std::abs(error.positions[i]) <= tolerances[i].position;
+        }
+        else return false;
```
Comment (fmauch) [inline pass_through_controllers/include/pass_through_controllers/pass_through_controllers.hpp]: the else can be skipped.
```suggestion
        return false;
```

File: pass_through_controllers/include/pass_through_controllers/pass_through_controllers.hpp [hunk 2]
```diff
@@ -243,28 +243,40 @@ namespace trajectory_controllers {
...
+      }
+
+      if (tolerances[i].velocity > 0.0)
       {
-        return false;
+        if (error.velocities.size() == tolerances.size())
+        {
+          return std::abs(error.velocities[i]) <= tolerances[i].velocity;
+        }
+        else return false;
```
Comment (fmauch) [inline pass_through_controllers/include/pass_through_controllers/pass_through_controllers.hpp]:
```suggestion
        return false;
```

File: pass_through_controllers/include/pass_through_controllers/pass_through_controllers.hpp [hunk 3]
```diff
@@ -243,28 +243,40 @@ namespace trajectory_controllers {
...
+      }
+
+      if (tolerances[i].acceleration > 0.0)
+      {
+        if (error.accelerations.size() == tolerances.size())
+        {
+          return std::abs(error.accelerations[i]) <= tolerances[i].acceleration;
+        }
+        else return false;
```
Comment (fmauch) [inline pass_through_controllers/include/pass_through_controllers/pass_through_controllers.hpp]:
```suggestion
        return false;
```

**Review by fmauch — COMMENTED** [review]
Hm. That implementation means that when I specify tolerances that the robot can't check, my trajectory will fail. As user it would be hard to find out what the problem actually is.

**Comment by stefanscherzinger** [comment]
> As user it would be hard to find out what the problem actually is.

That's true. We could think of adding additional error messages in this case.

> when I specify tolerances that the robot can't check, my trajectory will fail.

Exactly that's the intention. If it returned success, users might think that their tolerances were met..

**Comment by stefanscherzinger** [comment]
@fmauch We now provide an error message in 9be4ef0. What do you think?

File: pass_through_controllers/include/pass_through_controllers/pass_through_controllers.hpp [hunk 4]
```diff
@@ -241,36 +242,52 @@ namespace trajectory_controllers {
...
-           std::abs(error.velocities[i]) > tolerances[i].velocity) ||
-          (tolerances[i].acceleration > 0.0 &&
-           std::abs(error.accelerations[i]) > tolerances[i].acceleration))
+      if (tolerances[i].position > 0.0)
+      {
+        if (error.positions.size() == tolerances.size())
+        {
+          return std::abs(error.positions[i]) <= tolerances[i].position;
+        }
+        msg = "Position tolerances not fully supported by the driver implementation.";
```
Comment (fmauch) [inline pass_through_controllers/include/pass_through_controllers/pass_through_controllers.hpp]:
```suggestion
        msg = "Position tolerances specified, but not fully supported by the driver implementation.";
```

File: pass_through_controllers/include/pass_through_controllers/pass_through_controllers.hpp [hunk 5]
```diff
@@ -241,36 +242,52 @@ namespace trajectory_controllers {
...
+        return false;
+      }
+
+      if (tolerances[i].acceleration > 0.0)
+      {
+        if (error.accelerations.size() == tolerances.size())
+        {
+          return std::abs(error.accelerations[i]) <= tolerances[i].acceleration;
+        }
+        msg = "Acceleration tolerances not fully supported by the driver implementation.";
```
Comment (fmauch) [inline pass_through_controllers/include/pass_through_controllers/pass_through_controllers.hpp]:
```suggestion
        msg = "Acceleration tolerances  specified, but not fully supported by the driver implementation.";
```

File: pass_through_controllers/include/pass_through_controllers/pass_through_controllers.hpp [hunk 6]
```diff
@@ -241,36 +242,52 @@ namespace trajectory_controllers {
...
+        return false;
+      }
+
+      if (tolerances[i].velocity > 0.0)
       {
+        if (error.velocities.size() == tolerances.size())
+        {
+          return std::abs(error.velocities[i]) <= tolerances[i].velocity;
+        }
+        msg = "Velocity tolerances not fully supported by the driver implementation.";
```
Comment (fmauch) [inline pass_through_controllers/include/pass_through_controllers/pass_through_controllers.hpp]:
```suggestion
        msg = "Velocity tolerances specified, but not fully supported by the driver implementation.";
```

**Review by fmauch — COMMENTED** [review]
I'd suggest small modifications to the error message. Apart from that I like it!

**Comment by stefanscherzinger** [comment]
Thanks, that's better.

**Pull request merged**